### ISSUE TITLE
Quarkus 3.6.0 generate wrong jacoco xml

### ISSUE THREAD

**Issue by andrejpetras** [quarkusio/quarkus#37496]
### Describe the bug

After migration from Quarkus `3.5.3` to Quarkus `3.6.0`, the jacoco report file in our projects is not generated correctly. This happens only for an application which has multiple tests with custom/different test profiles.

Snippets from xml file:
```xml
    <counter type="COMPLEXITY" missed="0" covered="2"/>
    <counter type="METHOD" missed="0" covered="2"/>
    <counter type="CLASS" missed="0" covered="1"/>
</report>report>
```
In this example, the last line of the file is incorrect. Full example: https://github.com/andrejpetras/quarkus-ex-mapper/blob/feat/jacoco/examples/jacoco-3.xml#L46

```xml
                <counter type="LINE" missed="0" covered="1"/>
                <co1"/>
                <counter type="COMPLEXITY" missed="0" covered="1"/>
```

`  <co1"/>` Element?

```xml
        </sourcefile>
        <count="INSTRUCTION" missed="0" covered="8"/>
        <counter type="LINE" missed="0" covered="2"/>
```

`count` Element?

### Expected behavior

The jacoco xml report file is generated correctly as a valid xml format.

### Actual behavior

If the application does have multiple tests with different/custom test profiles (see example), the jacoco report xml file is not generated correctly.

Report examples: https://github.com/andrejpetras/quarkus-ex-mapper/tree/feat/jacoco/examples

### How to Reproduce?

I created an example application: https://github.com/andrejpetras/quarkus-ex-mapper/tree/feat/jacoco
To reproduce this issue run following steps:
1. build application `mvn clean package`
2. validate and check `target/jacoco-report/jacoco.xml` file.

### Output of `uname -a` or `ver`

Linux CE52819 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk 21.0.1 2023-10-17 LTS OpenJDK Runtime Environment Zulu21.30+15-CA (build 21.0.1+12-LTS) OpenJDK 64-Bit Server VM Zulu21.30+15-CA (build 21.0.1+12-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.6.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.5 (57804ffe001d7215b5e7bcb531cf83df38f93546) Maven home: /home/andrej/tools/maven Java version: 21.0.1, vendor: Azul Systems, Inc., runtime: /home/andrej/tools/zulu21.30.15-ca-jdk21.0.1-linux_x64 Default locale: en, platform encoding: UTF-8 OS name: "linux", version: "5.15.90.1-microsoft-standard-wsl2", arch: "amd64", family: "unix"

**Comment by gsmet** [comment]
Thanks for the report. 

It could be related to this jacoco upgrade? https://github.com/quarkusio/quarkus/pull/36648

Could you try to downgrade in your project and see how it goes?

If it fixes the issue, we can revert for 3.6.1 and you should open a Jacoco issue.

**Comment by famod** [comment]
This is rather flaky, so chances are you were just lucky you didn't hit it before 3.6.
I was able to make it fail after a few runs via:
- `curl https://raw.githubusercontent.com/jacoco/jacoco/master/org.jacoco.report/src/org/jacoco/report/xml/report.dtd > report.dtd`
- `while mvn clean package && cp report.dtd target/jacoco-report/ && xmllint --valid target/jacoco-report/jacoco.xml; do :; done`

To me the output looks like two threads are writing to the same file, maybe?
IIRC the report generation in Quarkus is rather fragile (due to some limitations imposed by JaCoCo).

I can try and have a closer look, but I don't know when this will be exactly.

**Comment by andrejpetras** [comment]
@gsmet I tried with Quarkus 3.6.0 and the old Jacoco version, but I got the same results.

I couldn't reproduce it in 3.5.3 version on multiple projects. But it could be that I was lucky.

**Comment by famod** [comment]
There we go: We register multiple shutdown hooks, one for the "unprofiled" test and one for the profiled test: https://github.com/quarkusio/quarkus/blob/3.6.1/test-framework/jacoco/deployment/src/main/java/io/quarkus/jacoco/deployment/JacocoProcessor.java#L118
Verified by adding some output.

I'll see what I can do.

**Comment by famod** [comment]
Fun fact: In 3.5.3 there are also two hook threads, but it just doesn't fail.

**Comment by famod** [comment]
I'll push a workaround tomorrow. It's far from pretty but it kinda serializes all hooks so that they don't write to the report files concurrently.

**Comment by bchetty** [comment]
And BTW, with this new version of Quarkus (3.6.0), the quarkus-jacoco generated report is also incorrect (with 0% coverage) for a maven-multi-module setup ( maven-surefire & maven-failsafe are configured as per the Quarkus documentation: https://quarkus.io/guides/tests-with-coverage )

**Comment by famod** [comment]
Does anyone want to give #37570 a try? Thanks!

**Comment by amkac** [comment]
> Thanks for the report.
> 
> It could be related to this jacoco upgrade? #36648
> 
> Could you try to downgrade in your project and see how it goes?
> 
> If it fixes the issue, we can revert for 3.6.1 and you should open a Jacoco issue.

@gsmet I git same issue with Quarkus 3.6.0 and Jacoco 0.8.10
